Initial design document

zship opened the pull request:
This is the first draft of my design document for the Teleport Frontend Application Developer Challenge.

zship committed:
Initial design document

kimlisa commented on design-doc/README.md:
```diff
@@ -0,0 +1,295 @@
+
+### `/filebrowser/:filepath`
+
+Components:
```
we also need a search functionality: search by filename or size

zship replied:
Ah! My mistake. I would normally implement search (and sort) on the client side unless pagination is needed. Seeing as this is a challenge, though, I'd understand wanting it to be implemented on the server side. Is it acceptable for me to add search into the "Client" part of this design doc, or would you/the team prefer the server side?

kimlisa replied:
adding search into the client is fine :)

zship replied:
Follow-up I just thought of! How is "search by size" intended to function? I can see two possibilities:
1. Lexical match, same as filename search, e.g. "32" matches "320 KB" but not "3.3 MB". This seems not super useful.
2. Conditional match, e.g. "size >= 10 KB". This would require me to specify several more form fields in the mockup.

kimlisa replied:
hm, as a user, i think i would expect behavior 1

zship replied:
Cool, thanks Lisa! I've updated the description and mockups to that effect.

kimlisa commented on design-doc/README.md:
```diff
@@ -0,0 +1,295 @@
+[cfwcpu]: https://www.npmjs.com/package/argon2
+[qdhapb]: https://www.npmjs.com/package/jsonwebtoken
+
+I'll also be including unit tests. Those will employ these development
```
Testing was stated under server, so making sure that we don't forget including tests for the react comps

zship replied:
Good point, I'll move this "unit tests" blurb up to the top where I talk about common development dependencies.

kimlisa commented on design-doc/README.md:
```diff
@@ -0,0 +1,295 @@
+[ydmgmg]: https://www.chaijs.com/
+
+
+## Endpoints
```
I don't think i saw handling for logout

zship replied:
You're right, I did leave that out!  Since I'm using short-lived access tokens with no refresh token, "logout" would just be a button/link on the client side that deletes the access token from localStorage.  Will add.

Edit: per [the conversation with Alexey below](https://github.com/zship/teleport-challenge/pull/1#discussion_r718764746), I've switched to traditional sessions and have implemented a logout endpoint and route.

alex-kovoy commented on design-doc/README.md:
```diff
@@ -0,0 +1,295 @@
+with a 1-hour expiration. In development mode a keypair will be generated for
+JWT signing.
+
+> Why JWTs? (tl;dr: my general desire to avoid state when possible) JWTs are a
```
I think that JWT makes it a bit more complicated as it will be harder to implement a proper logout and enforcing 1h session TTL seems as a bad UX. I would revisit this design choice as part of the challenge is to implement a logout mechanism.

zship replied:
That's fair! I went back-and-forth on this one myself. To get a good UX with JWTs I would need to implement refresh tokens, and then logout would require refresh token revocation lists, and then we're back to keeping track of state on the server.

I'll switch to using session IDs instead.

alex-kovoy commented on design-doc/README.md:
```diff
@@ -0,0 +1,295 @@
+such as "index.html", "index.js", "style.css", etc.
+
+A request for a directory or a non-existent file will return the contents of
+"index.html" to support reloading and bookmarking pushState-based client
```
Are you trying to say that all requests that do not match `/api` route will return an index.html? If so I would clarify this message as it seems that `index.html` will be returned for a directory path requests that does not exist.

alex-kovoy replied:
how redirects will be handled back to original URL (after a user has been successfully authenticated) ?

zship replied:
Yes, you're right that my intention is to return index.html for all routes not matching `/api`. I can see how that seems weird in this context! Normally I see single page applications like this implemented with two servers:

1. Static file server, usually nginx configured to serve index.html instead of 404s. This means the client is wholly responsible for routing.
2. API server

Here I'm combining the functionality of those two servers into one. Does that make sense? Perhaps it would be clearer to keep this functionality separated and either use a proxy server (to make them appear to the client to come from the same origin) or use CORS for the API server.

zship replied:
> how redirects will be handled back to original URL (after a user has been successfully authenticated) ?

Ah right, I'll add this to the "Client" -> `/login` section.

zship replied:
Added an explanation for login redirects.

r0mant commented on design-doc/README.md:
```diff
@@ -0,0 +1,295 @@
+
+Authentication and authorization will be handled using a bearer access token
+gotten from the server's `/login` endpoint. The access token will be stored in
+the browser's `localStorage` and passed in authenticated requests using the
```
Are there any downsides to using local storage for the access token vs cookies?

zship replied:
Technically yes, using localStorage means that the access token is able to be read in the event of an XSS attack.  A cookie (with httpOnly set) would be unreadable by malicious code. I might say it's about even, though, considering an XSS attack would still be able to make authenticated requests (just not read the access token directly).

Edit: would be interested to know if my understanding of this has gaps!

alex-kovoy commented on design-doc/README.md:
````diff
@@ -0,0 +1,295 @@
+
+```
+{
+  "code": "${errorCode}"
````
I would add an error message as well, this will greatly simplify client side error handling.

zship replied:
Good call out! I should probably also add HTTP response codes just to be explicit, come to think of it.

r0mant reviewed (approved):
lgtm but please wait for @kimlisa and @alex-kovoy approvals as well

alex-kovoy commented on design-doc/README.md:
```diff
@@ -0,0 +1,295 @@
+## Dependencies
+
+- React: required by the challenge (but I'd probably use it anyway)
+- Redux: not strictly required, but I like its reducer pattern and I think the
```
What are the advantages of using `Redux` for this app?

zship replied:
I mainly like the Redux DevTools extension, particularly for its logging of every event that could have caused a component to re-render.  It can sometimes be easier for me to notice why a component is rendering in an unexpected way and/or decide where to place breakpoints if I have the ability to first step *backwards* through a full log of changes.

That said, this is a very small project with a single developer so debugging is already quite a bit easier. I'm open to trying a different way of doing things.

alex-kovoy replied:
Considering that Redux is loosing it's popularity (mainly because after React Context API and hooks being released) I suggest to stick with built-in state management.

zship replied:
Fair enough, removed.

alex-kovoy commented on design-doc/README.md:
```diff
@@ -0,0 +1,295 @@
+
+- React: required by the challenge (but I'd probably use it anyway)
+- Redux: not strictly required, but I like its reducer pattern and I think the
+  Redux DevTools extension is helpful in development
```
How will the routes be handled?

zship replied:
I like [this approach](https://medium.com/@daveford/react-router-alternative-switch-acd7961f08db) personally (a function that parses the URL and returns a React element). I think it's a bit easier to understand than React Router. There's something to be said for using well-known libraries, though, so I'd be happy to use React Router instead if that's a preference of the team.

alex-kovoy commented on design-doc/README.md:
```diff
@@ -0,0 +1,295 @@
+  incorrect
+
+
+### `GET /api/filebrowser/:filepath`
```
Are there going to be any input validation regarding these requests?

zship replied:
Yes, good point! I could probably be a bit more explicit in this document about expected data types.

zship replied:
I've added data types and an (optional/required) note to each request parameter. Let me know if you were looking for anything else.

alex-kovoy replied:
I was looking at something like the following 

```
/dir/../../../`
```

 or say a couple of words about encoding/decoding (if you are not planning to use react router for that)

zship replied:
Oh boy, that went right over my head. I mentioned a similar issue [here](https://github.com/zship/teleport-challenge/blob/eb2b9010d917b6ab86c8e85001e8386190225a6b/design-doc/README.md#L305) but didn't think about a relative path on the `GET /api/filebrowser` endpoint because it's not dealing with a real filesystem (it's using a fake JSON directory structure). But conceptually this endpoint should be treated as if it's a real FS, you're right. I'll document my planned implementation here.

zship replied:
Added some discussion of `filepath` handling for both the [client](https://github.com/zship/teleport-challenge/blob/812471ee12663b0a89ba416fa458b25c3b56e6e9/design-doc/README.md#L108) and [server](https://github.com/zship/teleport-challenge/blob/812471ee12663b0a89ba416fa458b25c3b56e6e9/design-doc/README.md#L256-L265).

alex-kovoy replied:
as per service requirements (if i am not mistaken), a read directory needs to be used of JSON.

zship replied:
Got it, `GET /api/filebrowser` should operate against a real filesystem. I updated this section accordingly.

zship committed:
PR #1 feedback:

- Switched auth method from JWT to session-based
- Added `/api/logout` endpoint and `/logout` client route
- Added user-friendly error messages to HTTP error responses along with
  HTTP status codes
- Removed Redux dependency

zship committed:
PR #1 feedback: moved the "unit test" statement from the "Server" section to the top (to apply to both "Server" and "Client")

zship committed:
Copy/paste mistake: corrected the request body definition for /api/logout

zship committed:
PR #1 feedback: added expected validation for the /api/login endpoint

zship committed:
PR #1 feedback: added an explanation for login redirect

zship committed:
PR #1 feedback: added search function to DataGrid (a sub-component of FileBrowser)

russjones commented on design-doc/README.md:
```diff
@@ -0,0 +1,406 @@
+
+CSRF attacks are not a concern for two reasons:
+
+- There are no HTTP requests with side-effects, i.e. the server is stateless.
```
Doesn't the login endpoint change state?

zship replied:
Ah! Yes, good spot. Login was stateless when I wrote that, but not now. Will remove.

zship replied:
Added some discussion [here](https://github.com/zship/teleport-challenge/blob/eb2b9010d917b6ab86c8e85001e8386190225a6b/design-doc/README.md#L75)

russjones commented on design-doc/README.md:
```diff
@@ -0,0 +1,406 @@
+<img src="./assets/DataGrid-search-size.png" />
+
+
+## Security Considerations
```
What about open redirect?

zship replied:
Interesting, I wasn't aware of that one! The `/login?redirect=` route should be rethought. Will do.

zship replied:
Added some discussion [here](https://github.com/zship/teleport-challenge/blob/eb2b9010d917b6ab86c8e85001e8386190225a6b/design-doc/README.md#L75)

russjones commented on design-doc/README.md:
````diff
@@ -0,0 +1,406 @@
+]
+```
+
+Passwords will be hashed as follows:
+
+1. The plaintext password will be hashed using SHA512. [This is a technique
+   used by Dropbox to help defend against DoS attacks.][atpftb]
+2. A random salt will be generated per-user. Length will be 128 bits (arbitrary
+   choice; in a real application an optimal salt length could be computed using
+   a [birthday problem approximation][jwgsnw]).
+3. The SHA512 hash will be hashed again using **argon2id** with the random salt
+   and parameters `m=37 MiB, t=1, p=1`, per current [OWASP
+   recommendations][cjcbmz].
+4. The resulting argon2id hash will be encrypted with an AES256
+   [pepper][svzrcp].
````
You don't need to do all of this for this challenge, just using a KDF like bcrypt is fine.

zship replied:
Good to know! I'll go for scrypt in that case (it's already built in to node).

alex-kovoy commented on design-doc/README.md:
```diff
@@ -0,0 +1,406 @@
+  login. This is to support bookmarking authenticated paths. Intended to be
+  passed in from the router, extracted from a `window.location` that looks
+  like: `/login?redirect=<url-encoded-path>`, e.g.
+  `/login?redirect=%2Ffilebrowser%2Fteleport%2Flib`.
```
hm...why `/` need to be encoded in the `redirect` parameter?

zship replied:
URL encoding here is to make future query string parameters easier to parse. Not that there will be any in this particular case... hmm.

zship committed:
Address PR feedback:
- removed claim that the server is stateless (login is stateful)
- documented open redirect defense for `/login?redirect=` route
- simplified password hashing

zship committed:
Addressed PR feedback: documented `filepath` format in both client (/filebrowser/:filepath) and server side (/api/filebrowser/:filepath)

zship committed:
Made `redirect` parameter documentation slightly clearer

zship committed:
Addressed PR feedback: /api/filebrowser should operate against a real filesystem, so discuss how to prevent users breaking out of the allowed directory
